Common/Securities/CurrencyConversion/SecurityCurrencyConversion.cs

```diff
@@ -0,0 +1,268 @@
+
+            /// <summary>
+            /// Whether the price of the security must be inverted in the conversion
+            /// </summary>
+            public bool Inverted { get; }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Step"/> class
+            /// </summary>
+            /// <param name="rateSecurity">The security to use in this currency conversion step</param>
+            /// <param name="inverted">Whether the price of the security should be inverted in the conversion</param>
+            public Step(Security rateSecurity, bool inverted)
+            {
+                RateSecurity = rateSecurity;
+                Inverted = inverted;
+            }
+        }
+
+        private readonly List<Step> _steps;
+
+        /// <summary>
+        /// The currency this conversion converts from
+        /// </summary>
+        public string SourceCurrency { get; }
+
+        /// <summary>
+        /// The currency this conversion converts to
+        /// </summary>
+        public string DestinationCurrency { get; }
+
+        /// <summary>
+        /// The current conversion rate
+        /// </summary>
+        public decimal ConversionRate { get; private set; }
+
+        /// <summary>
+        /// The securities which the conversion rate is based on
+        /// </summary>
+        public IEnumerable<Security> ConversionRateSecurities => _steps.Select(step => step.RateSecurity);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SecurityCurrencyConversion"/> class.
+        /// This constructor is intentionally private as only <see cref="LinearSearch"/> is supposed to create it.
+        /// </summary>
+        /// <param name="sourceCurrency">The currency this conversion converts from</param>
+        /// <param name="destinationCurrency">The currency this conversion converts to</param>
+        /// <param name="steps">The steps between sourceCurrency and destinationCurrency</param>
+        private SecurityCurrencyConversion(string sourceCurrency, string destinationCurrency, List<Step> steps)
+        {
+            SourceCurrency = sourceCurrency;
+            DestinationCurrency = destinationCurrency;
+
+            _steps = steps;
+        }
+
+        /// <summary>
+        /// Updates the internal conversion rate based on the latest data, and returns the new conversion rate
+        /// </summary>
+        /// <returns>The new conversion rate</returns>
+        public decimal Update()
+        {
+            var newConversionRate = 0m;
+            var stepWithDataFound = false;
+
+            _steps.ForEach(step =>
+            {
+                var lastData = step.RateSecurity.GetLastData();
+                if (lastData == null)
+                {
+                    return;
+                }
+
+                var price = lastData.Price;
+                if (price == 0m)
+                {
+                    return;
+                }
```

Comment:
If one of the rate securities does not have data and one does, what would happen? A nice unit test to add, I think it should keep a 0 value probably

Reply:
Changed it to use a conversion rate of 0 if data for any of the steps is missing